Source/Schema.NET/OneOrMany{T}.cs

```diff
@@ -15,131 +16,130 @@ namespace Schema.NET
-                this.collection = items;
-                this.item = default;
+                if (index > 0)
+                {
+                    if (index == array.Length)
+                    {
+                        this.collection = items;
+                    }
+                    else
+                    {
+                        this.collection = new T[index];
+                        Array.Copy(items, 0, this.collection, 0, index);
+                    }
+
+                    this.HasOne = index == 1;
+                    return;
+                }
             }
+
+            this.collection = null;
+            this.HasOne = false;
         }
 
         /// <summary>
-        /// Gets the number of elements contained in the <see cref="OneOrMany{T}"/>.
+        /// Initializes a new instance of the <see cref="OneOrMany{T}"/> struct.
         /// </summary>
-        public int Count
+        /// <param name="collection">The collection of values.</param>
+        public OneOrMany(IEnumerable<T> collection)
+            : this(collection.ToArray())
         {
-            get
-            {
-                if (this.HasOne)
-                {
-                    return 1;
-                }
-                else if (this.HasMany)
-                {
-                    return this.collection.Count;
-                }
+        }
 
-                return 0;
-            }
+        /// <summary>
+        /// Initializes a new instance of the <see cref="OneOrMany{T}"/> struct.
+        /// </summary>
+        /// <param name="collection">The list of values.</param>
+        public OneOrMany(IEnumerable<object> collection)
+            : this(collection?.Cast<T>())
+        {
         }
 
+        /// <summary>
+        /// Gets the number of elements contained in the <see cref="OneOrMany{T}"/>.
+        /// </summary>
+        public int Count => this.collection?.Length ?? 0;
+
         /// <summary>
         /// Gets a value indicating whether this instance has a single item value.
         /// </summary>
         /// <value><c>true</c> if this instance has a single item value; otherwise, <c>false</c>.</value>
-        public bool HasOne => this.collection is null && this.item != null;
+        public bool HasOne { get; }
 
         /// <summary>
         /// Gets a value indicating whether this instance has more than one value.
         /// </summary>
         /// <value><c>true</c> if this instance has more than one value; otherwise, <c>false</c>.</value>
-        public bool HasMany => this.collection != null;
+        public bool HasMany => this.collection?.Length > 1;
 
         /// <summary>
         /// Performs an implicit conversion from <typeparamref name="T"/> to <see cref="OneOrMany{T}"/>.
         /// </summary>
         /// <param name="item">The single item value.</param>
         /// <returns>The result of the conversion.</returns>
 #pragma warning disable CA2225 // Operator overloads have named alternates

```

Comment:
What do you think of implementing the methods to do away with these warnings?

Reply:
Yeah, implementing those methods shouldn't be an issue. [Looking up the docs](https://docs.microsoft.com/en-us/visualstudio/code-quality/ca2225?view=vs-2019), it sounds like we just need `FromX` methods though it is vague whether those are meant to be static or what we would call it in the case of `T` - `FromT`?

Reply:
Going over this one again - I'm not sure we can implement good names that satisfy the rules.

For `T[]`, it wants the method named either `To` or `FromOneOrMany`. The former is just bad while the latter seems to imply a method on `T[]` itself - I don't think extension methods count to this analyzer rule.

For `T`, it wants the method named either `ToT` or `FromOneOrMany`. Now the former is slightly better than it was above but the latter will now be in conflict with above as two methods can't only be different by their return type.

When casting from a `List<T>` to `OneOrMany<T>`, we hit another problem. To have a method actually be `FromList`, it would make sense to be static on `OneOrMany<T>`. There is a conflicting rule that says static methods should not be on generic types. We could have `FromList` as a non-static method but that means we need a `OneOrMany<T>` to begin with.

I'm not sure what we want to do to satisfy the analyzer's specific naming requirements. In the mean time though, I did add decently named methods `ToArray` and `FirstOrDefault` which is what two of the implicit conversion use. While named like their LINQ counterparts, they are tailored for our data structure.

Reply:
That must be why I didn't do it in the first place then. I guess we'll leave it as is then?